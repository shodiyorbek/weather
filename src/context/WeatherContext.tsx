import { createContext, useContext, useReducer, useCallback, type ReactNode } from 'react';
//@ts-ignore
import type { WeatherState, WeatherAction } from '../types/weather';
import { weatherReducer, initialState } from '../lib/weatherReducer';

interface WeatherContextType {
  state: WeatherState;
  dispatch: React.Dispatch<WeatherAction>;
}

const WeatherContext = createContext<WeatherContextType | undefined>(undefined);

export function WeatherProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(weatherReducer, initialState);

  // Wrap dispatch with logging
  const wrappedDispatch = useCallback((action: WeatherAction) => {
    console.log('Dispatching action:', action);
    dispatch(action);
  }, []);

  return (
    <WeatherContext.Provider value={{ state, dispatch: wrappedDispatch }}>
      {children}
    </WeatherContext.Provider>
  );
}

export function useWeather() {
  const context = useContext(WeatherContext);
  if (context === undefined) {
    throw new Error('useWeather must be used within a WeatherProvider');
  }
  return context;
} 