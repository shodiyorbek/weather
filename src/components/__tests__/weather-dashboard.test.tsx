import { render, screen, waitFor } from '@testing-library/react'
import { describe, it, expect, vi } from 'vitest'
import { act } from 'react-dom/test-utils'
import WeatherDashboard from '../weather-dashboard'
import { WeatherProvider } from '@/context/WeatherContext'

// Mock the hooks and components
vi.mock('@/hooks/useWeatherData', () => ({
  useWeatherData: () => ({
    forecast: [],
    hourlyForecast: [],
    isLoading: false,
    lastUpdated: new Date('2024-01-01'),
    timeZone: 'UTC',
    isDetectingLocation: false,
    loadWeatherData: vi.fn(),
    detectUserLocation: vi.fn(),
    handleDetectLocation: vi.fn(),
    handleRefresh: vi.fn()
  })
}))

vi.mock('@/hooks/useFavoriteCities', () => ({
  useFavoriteCities: () => ({
    favoriteCities: [],
    addFavoriteCity: vi.fn(),
    removeFavoriteCity: vi.fn(),
    isFavoriteCity: vi.fn()
  })
}))

// Mock child components
vi.mock('@/components/dynamic-greeting', () => ({
  default: () => <div data-testid="mock-dynamic-greeting">Dynamic Greeting</div>
}))

vi.mock('@/components/weather-hero', () => ({
  default: () => <div data-testid="mock-weather-hero">Weather Hero</div>
}))

vi.mock('@/components/weather-stats', () => ({
  default: () => <div data-testid="mock-weather-stats">Weather Stats</div>
}))

vi.mock('@/components/hourly-forecast', () => ({
  default: () => <div data-testid="mock-hourly-forecast">Hourly Forecast</div>
}))

vi.mock('@/components/hourly-weather-chart', () => ({
  default: () => <div data-testid="mock-hourly-weather-chart">Hourly Weather Chart</div>
}))

vi.mock('@/components/weekly-forecast', () => ({
  default: () => <div data-testid="mock-weekly-forecast">Weekly Forecast</div>
}))

vi.mock('@/components/weather-chart', () => ({
  default: () => <div data-testid="mock-weather-chart">Weather Chart</div>
}))

vi.mock('@/components/settings-panel', () => ({
  default: () => <div data-testid="mock-settings-panel">Settings Panel</div>
}))

vi.mock('@/components/site-tour', () => ({
  default: () => <div data-testid="mock-site-tour">Site Tour</div>
}))

vi.mock('@/components/search-bar', () => ({
  default: () => <div data-testid="mock-search-bar">Search Bar</div>
}))

vi.mock('@/components/error-boundry', () => ({
  default: () => <div data-testid="mock-error-boundary">Error Boundary</div>
}))

describe('WeatherDashboard', () => {
  it('should match snapshot', () => {
    const { container } = render(
        <WeatherProvider>
          <WeatherDashboard />
        </WeatherProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render all main components in overview view', () => {
    render(
        <WeatherProvider>
          <WeatherDashboard />
        </WeatherProvider>
    )

    // Check for components that are always visible
    expect(screen.getByTestId('mock-dynamic-greeting')).toBeInTheDocument()
    expect(screen.getByTestId('mock-weather-hero')).toBeInTheDocument()
    expect(screen.getByTestId('mock-weather-stats')).toBeInTheDocument()
    expect(screen.getByTestId('mock-hourly-weather-chart')).toBeInTheDocument()
    expect(screen.getByTestId('mock-weekly-forecast')).toBeInTheDocument()
    expect(screen.getByTestId('mock-weather-chart')).toBeInTheDocument()
    expect(screen.getByTestId('mock-site-tour')).toBeInTheDocument()
    expect(screen.getByTestId('mock-search-bar')).toBeInTheDocument()
    expect(screen.getByTestId('mock-error-boundary')).toBeInTheDocument()
  })

  it('should render hourly view components', async () => {
    render(
        <WeatherProvider>
          <WeatherDashboard />
        </WeatherProvider>
    )

    // Click the forecast tab
    const forecastTab = screen.getByRole('button', { name: /forecast/i })
    await act(async () => {
      forecastTab.click()
    })

    // Wait for the components to be rendered
    await waitFor(() => {
      expect(screen.getByTestId('mock-hourly-forecast')).toBeInTheDocument()
      expect(screen.getByTestId('mock-weekly-forecast')).toBeInTheDocument()
    })
  })

  it('should render weekly view components', async () => {
    render(
        <WeatherProvider>
          <WeatherDashboard />
        </WeatherProvider>
    )

    // Click the statistic tab
    const statisticTab = screen.getByRole('button', { name: /statistic/i })
    await act(async () => {
      statisticTab.click()
    })

    // Wait for the components to be rendered
    await waitFor(() => {
      expect(screen.getByTestId('mock-weekly-forecast')).toBeInTheDocument()
    })
  })
}) 