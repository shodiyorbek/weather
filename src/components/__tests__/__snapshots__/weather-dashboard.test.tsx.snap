// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeatherDashboard > should match snapshot 1`] = `
<div>
  <div
    class="min-h-screen bg-gradient-to-br from-blue-50 via-indigo-50 to-purple-50 dark:from-slate-900 dark:via-blue-900 dark:to-indigo-900 transition-all duration-500"
  >
    <div
      data-testid="mock-site-tour"
    >
      Site Tour
    </div>
    <div
      class="absolute inset-0 opacity-50"
    />
    <div
      class="relative z-10"
    >
      <header
        class="sticky top-0 z-50 backdrop-blur-xl bg-white/80 dark:bg-slate-900/80 border-b border-white/20 dark:border-slate-700/50"
      >
        <div
          class="container mx-auto px-4 py-4"
        >
          <div
            class="flex flex-col md:flex-row justify-between items-start lg:items-center gap-4"
          >
            <div
              class="flex items-center gap-6 justify-between md:w-fit w-full"
            >
              <div
                class="block"
              >
                <div
                  data-testid="mock-dynamic-greeting"
                >
                  Dynamic Greeting
                </div>
              </div>
              <div
                class="md:hidden flex"
              >
                <div
                  class="flex items-center gap-2"
                >
                  <button
                    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 location-button bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                    data-slot="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="lucide lucide-navigation h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <polygon
                        points="3 11 22 2 13 21 11 13 3 11"
                      />
                    </svg>
                  </button>
                  <button
                    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                    data-slot="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="lucide lucide-refresh-cw h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
                      />
                      <path
                        d="M21 3v5h-5"
                      />
                      <path
                        d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
                      />
                      <path
                        d="M8 16H3v5"
                      />
                    </svg>
                  </button>
                  <button
                    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 settings-button bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                    data-slot="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="lucide lucide-settings h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12.22 2h-.44a2 2 0 0 0-2 2v.18a2 2 0 0 1-1 1.73l-.43.25a2 2 0 0 1-2 0l-.15-.08a2 2 0 0 0-2.73.73l-.22.38a2 2 0 0 0 .73 2.73l.15.1a2 2 0 0 1 1 1.72v.51a2 2 0 0 1-1 1.74l-.15.09a2 2 0 0 0-.73 2.73l.22.38a2 2 0 0 0 2.73.73l.15-.08a2 2 0 0 1 2 0l.43.25a2 2 0 0 1 1 1.73V20a2 2 0 0 0 2 2h.44a2 2 0 0 0 2-2v-.18a2 2 0 0 1 1-1.73l.43-.25a2 2 0 0 1 2 0l.15.08a2 2 0 0 0 2.73-.73l.22-.39a2 2 0 0 0-.73-2.73l-.15-.08a2 2 0 0 1-1-1.74v-.5a2 2 0 0 1 1-1.74l.15-.09a2 2 0 0 0 .73-2.73l-.22-.38a2 2 0 0 0-2.73-.73l-.15.08a2 2 0 0 1-2 0l-.43-.25a2 2 0 0 1-1-1.73V4a2 2 0 0 0-2-2z"
                      />
                      <circle
                        cx="12"
                        cy="12"
                        r="3"
                      />
                    </svg>
                  </button>
                  <button
                    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                    data-slot="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="lucide lucide-moon h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
            <div
              data-testid="mock-search-bar"
            >
              Search Bar
            </div>
            <div
              class="items-center gap-2 hidden md:flex"
            >
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 location-button bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                data-slot="button"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-navigation h-4 w-4"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="3 11 22 2 13 21 11 13 3 11"
                  />
                </svg>
              </button>
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                data-slot="button"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-refresh-cw h-4 w-4"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
                  />
                  <path
                    d="M21 3v5h-5"
                  />
                  <path
                    d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
                  />
                  <path
                    d="M8 16H3v5"
                  />
                </svg>
              </button>
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 settings-button bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                data-slot="button"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-settings h-4 w-4"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12.22 2h-.44a2 2 0 0 0-2 2v.18a2 2 0 0 1-1 1.73l-.43.25a2 2 0 0 1-2 0l-.15-.08a2 2 0 0 0-2.73.73l-.22.38a2 2 0 0 0 .73 2.73l.15.1a2 2 0 0 1 1 1.72v.51a2 2 0 0 1-1 1.74l-.15.09a2 2 0 0 0-.73 2.73l.22.38a2 2 0 0 0 2.73.73l.15-.08a2 2 0 0 1 2 0l.43.25a2 2 0 0 1 1 1.73V20a2 2 0 0 0 2 2h.44a2 2 0 0 0 2-2v-.18a2 2 0 0 1 1-1.73l.43-.25a2 2 0 0 1 2 0l.15.08a2 2 0 0 0 2.73-.73l.22-.39a2 2 0 0 0-.73-2.73l-.15-.08a2 2 0 0 1-1-1.74v-.5a2 2 0 0 1 1-1.74l.15-.09a2 2 0 0 0 .73-2.73l-.22-.38a2 2 0 0 0-2.73-.73l-.15.08a2 2 0 0 1-2 0l-.43-.25a2 2 0 0 1-1-1.73V4a2 2 0 0 0-2-2z"
                  />
                  <circle
                    cx="12"
                    cy="12"
                    r="3"
                  />
                </svg>
              </button>
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 bg-white/70 dark:bg-slate-800/70 backdrop-blur-sm hover:bg-white/90 dark:hover:bg-slate-700/90"
                data-slot="button"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-moon h-4 w-4"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="flex items-center gap-1 mt-4 p-1 bg-white/50 dark:bg-slate-800/50 rounded-xl backdrop-blur-sm"
          >
            <button
              class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 flex-1 bg-white dark:bg-slate-700 text-black hover:bg-black/10 shadow-sm"
              data-slot="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-cloud h-4 w-4 mr-2"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17.5 19H9a7 7 0 1 1 6.71-9h1.79a4.5 4.5 0 1 1 0 9Z"
                />
              </svg>
              Current weather
            </button>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 flex-1 hover:bg-white/50 dark:hover:bg-slate-700/50"
              data-slot="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-clock h-4 w-4 mr-2"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <polyline
                  points="12 6 12 12 16 14"
                />
              </svg>
              Forecast
            </button>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 flex-1 hover:bg-white/50 dark:hover:bg-slate-700/50"
              data-slot="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-moon h-4 w-4 mr-2"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
                />
              </svg>
              Statistic
            </button>
          </div>
        </div>
      </header>
      <div
        data-testid="mock-error-boundary"
      >
        Error Boundary
      </div>
      <main
        class="container mx-auto px-4 py-6"
      >
        <div
          class="space-y-8"
          style="opacity: 0; transform: translateY(20px);"
        >
          <div
            class="flex items-center gap-2 overflow-x-auto pb-2 favorites-section"
          >
            <span
              class="text-sm font-medium text-slate-600 dark:text-slate-400 whitespace-nowrap"
            >
              Favorites:
            </span>
          </div>
          <div
            class="weather-hero"
          >
            <div
              data-testid="mock-weather-hero"
            >
              Weather Hero
            </div>
          </div>
          <div
            class="hourly-forecast"
          >
            <h3
              class="text-xl font-semibold mb-4 text-slate-800 dark:text-white flex items-center gap-2"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-clock h-5 w-5 text-blue-500"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <polyline
                  points="12 6 12 12 16 14"
                />
              </svg>
              Hourly Forecast
            </h3>
            <div
              data-testid="mock-hourly-weather-chart"
            >
              Hourly Weather Chart
            </div>
          </div>
          <div
            class="weekly-forecast"
          >
            <div
              data-testid="mock-weekly-forecast"
            >
              Weekly Forecast
            </div>
          </div>
          <div
            class="flex flex-col gap-6"
          >
            <div
              data-testid="mock-weather-stats"
            >
              Weather Stats
            </div>
            <div
              data-testid="mock-weather-chart"
            >
              Weather Chart
            </div>
          </div>
        </div>
      </main>
      <footer
        class="mt-12 py-6 text-center text-sm text-slate-500 dark:text-slate-400"
      >
        <p>
          Last updated: 
          1/1/2024, 5:00:00 AM
           • Data provided by OpenWeatherMap
        </p>
      </footer>
    </div>
  </div>
</div>
`;
